release/authorization/authorization.proto

```diff
@@ -0,0 +1,115 @@
+// Authorization for gnXi operations.
+//
+// Define a list of policies, where the most specific policy is applied
+// on the device authorizing the action.
+//
+// Users may be referenced by user name, or as a group with a group name.
+// Groups are defined and managed in the authorization policy itself.
+//
+// Paths may be referenced in whole or in complete parts, ie:
+//   /interfaces/interface[name=Ethernet1/2/3]/state/counters
+//   /interfaces/interface[name=*]/state/oper-status
+//   /network-instances/network-instance/tables/table[protocol=BGP][address-family=*]
+//
+// Paths are gnmi.Path protobufs.
+//
+// The most specific match is returned for a request, this means the longest
+// path and the most specific user match (user preferred over group).
+//
+// Additionally, the policy maintains a timestamp of creation
+// and version number from the underlying version control system.
+// The version/timestamp are available to requestors to verify which version
+// of policy is being applied at the time of the request.
+//
+syntax = "proto3";
+
+package authorization.proto;
+
+import "third_party/openconfig/gnmi/proto/gnmi/gnmi.proto";
+
+option java_multiple_files = true;
+option go_api_flag = "OPAQUE_V0";
+
+// User is a singular username used only in the matching criteria.
+message User {
+  string name = 1;
+}
+
+// Group is a group of users, groups are a construct of the policy configuration
+// as a method to collect many users with the same authorizations together.
+message Group {
+  string name = 1;
+  repeated User members = 2;
+}
+
+// LogLevel is set for all RPCs (via default_log_level, or per-RPC specifically.
+enum LogLevel {
+  LOG_NONE = 0;   // Don't pollute the logs with high volume RPCs.
+  LOG_BRIEF = 1;  // RPC name and calling Principal name.
+  LOG_FULL = 2;   // RPC, Principal, Request, and Response.
+}
+
+// Action is the defined action for an AuthorizationPolicy.
+enum Action {
+  UNKNOWN = 0;
+  DENY = 1;
+  PERMIT = 2;
+}
+
+message AuthorizationPolicy {
+  // Policy Identifier, a unique string per policy/rule, eg: uuid.
+  string id = 1;
+  LogLevel log_level = 2;
+
+  oneof principal {
+    User user = 3;
+    Group group = 4;
```

Comment:
Since these use the message types directly, this means that each principal within a policy contains the entire message construct each time.  For the case of a `Group`, while you can define these at the upper message layer of `AuthorizationPolicies`, the best reuse would be to only reference the User.name and Group.name only (especially if they are ever expanded with additional attributes)

Reply:
(I have no idea how to quote the original conversation here... either github sucks or I suck at github, I'm leaning on 2nd)

Eben: Agreed about the need for a RO/RW setting on the authorization policy.

For your second comment/suggestion i'm not sure I follow. As I was thinking about this originally the specification of a user or group in a policy (this policy applies to USER-X or GROUP-Y) was meant as a method to either: 
  "All of these users have this permission/policy"
or:
  "This one user has a specific policy/permission"

There's not a constraint on whether the username / groupname must be unique (between the two), though there could be. I'm probably misunderstanding your comment, however.

Reply:
fyi: Added the RO/RW bits to the proto, as a new ENUM and as a message element in  the AuthorizationPolicy message.

Reply:
The second comment means to use "string group" instead of "Group group" IIUC. Technically, we can just fill the group name in "Group" here. But changing the type to string makes it more clear.

Reply:
@mint570 ok, how would you differentiate the policy for
  user: Robert Oot - root
  group: root

or any other clash in user/group name uniqueness?
I didn't want to force the creator of the policy to have to make user/group names unique.
I suppose we COULD do that, it's not totally unreasonable, but it's not what I was thinking :)

Reply:
The proto field name tells it is a user or a group.
BTW, shouldn't user and group be a repeated field? We can have multiple users and groups on the same path.

Reply:
@mint570 I'm ready to believe you, so compiling the current proto:
  protoc --proto_path=. --go_out=. --go_opt=paths=source_relative  authorization.proto 

Gets me this snippet:
  https://gist.github.com/morrowc/542b0eacf4b7f3286210523793c28acd

I believe this code snippet, which did work before sending this PR, extracts the group/user content from a policy:
  https://gist.github.com/morrowc/d46fdc063eedb6cedf9bc77f75019ea3

This just, really, checks first:
  "Is there a user in the policy? and is that user the same as that which the policy evaluation is being performed for?"

Swapping the User/Group enum for just a string:
  https://gist.github.com/morrowc/022f4bfb2d8fda68019d330f5c03a5c5

The compiled bits are:
  https://gist.github.com/morrowc/e837b30e6713386d9dc5202dee542444

This seems like it will work out in the same manner in the code above, minus requesting the .GetName().
Cool, probably :) (see updated commit and thanks!)

For the repeated field conversation, internally I think the answer was really:
  It's simpler to do this as:
     "If you have many users with the same policy, make a group"
     "If you have many groups that need the same policy, make a group"
  in places where we've intentionally grouped things:
    "joe, jane, sam, sally all need this access"
  we've invariably ended up needing to split these as we add/remove access for
  one/some of the users in the list... Often this happens after we remove acesss for userX
  promptly breaking userY along the way.

Effectively it might be a little more verbose, but leaves the door open for easy and clear flexibility.